src/data/list/dedup.lean

```diff
@@ -77,4 +77,42 @@ lemma repeat_dedup {x : α} : ∀ {k}, k ≠ 0 → (repeat x k).dedup = [x]
 | (n+2) _ := by rw [repeat_succ, dedup_cons_of_mem (mem_repeat.2 ⟨n.succ_ne_zero, rfl⟩),
     repeat_dedup n.succ_ne_zero]
 
+lemma count_dedup (l : list α) (a : α) :
+  l.dedup.count a = if a ∈ l then 1 else 0 :=
+begin
+  refine trans (count_eq_of_nodup $ nodup_dedup l) _,
+  simp_rw [mem_dedup],
+end
+
+/-- Summing the count of `x` over a list filtered by some `p` is just `countp` applied to `p` -/
+lemma sum_map_count_dedup_filter_eq_countp (p : α → Prop) [decidable_pred p]
+  (l : list α) : ((l.dedup.filter p).map $ λ x, l.count x).sum = l.countp p :=
+begin
+  induction l with a as h,
+  { simp },
+  { simp_rw [list.countp_cons, list.count_cons', list.sum_map_add],
+    congr' 1,
+    { refine trans _ h,
+      by_cases ha : a ∈ as,
+      { simp [dedup_cons_of_mem ha] },
+      { simp only [dedup_cons_of_not_mem ha, list.filter],
+        split_ifs with hp; simp [list.map_cons, list.sum_cons,
+          list.count_eq_zero.2 ha, zero_add] } },
+    { by_cases hp : p a,
+      { refine trans (sum_map_eq_nsmul_single a _ (λ _ h _, by simp [h])) _,
+        simp [hp, count_dedup] },
+      { refine trans (list.sum_eq_zero $ λ n hn, _) (by simp [hp]),
+        obtain ⟨a', ha'⟩ := list.mem_map.1 hn,
+        simp only [(λ h, hp (h ▸ (list.mem_filter.1 ha'.1).2) : a' ≠ a), if_false] at ha',
+        exact ha'.2.symm } } },
+end
+
+lemma sum_map_count_dedup_eq_length (l : list α) :
+  (l.dedup.map $ λ x, l.count x).sum = l.length :=
+calc (l.dedup.map $ λ x, l.count x).sum
+  = ((l.dedup.filter $ λ x, true).map (λ x, l.count x)).sum :
+    by { congr, exact symm (filter_true _) }
+  ... = l.countp (λ x, true) : sum_map_count_dedup_filter_eq_countp _ l
+  ... = l.length : (countp_eq_length _).2 (by simp)
```

Comment:
Since I don't think this proof is mathematically interesting
```suggestion
by simpa using sum_map_count_dedup_filter_eq_countp (λ _, true) l
```
which nicely reads as "this follows trivially from the previous result"